### PR TITLE
Migrate virus check job to sidekiq

### PR DESCRIPTION
**Pull request by chrislo**
Contributes to: #179 

This PR migrates the virus scan delayed job worker to sidekiq. It leaves `Asset#scan_for_viruses` in place in case there are still jobs in delayed_job that expect that method to be present. This can be removed in future when we are sure there are no more jobs being added to the delay job queue.

### CODE REVIEW

File: .rubocop_todo.yml [hunk 1]
```diff
@@ -68,5 +69,6 @@ RSpec/VerifiedDoubles:
     - 'spec/controllers/media_controller_spec.rb'
     - 'spec/lib/virus_scanner_spec.rb'
     - 'spec/models/asset_spec.rb'
+    - 'spec/workers/virus_scan_worker_spec.rb'
```
Comment (floehopper) [inline .rubocop_todo.yml]: Very minor: Do we definitely still need the exceptions for `spec/models/asset_spec.rb` in these two lists? i.e. are there still Rubocop violations in the `Asset` spec?

File: spec/workers/virus_scan_worker_spec.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
+require 'rails_helper'
+
+RSpec.describe VirusScanWorker, type: :worker do
```
Comment (floehopper) [inline spec/workers/virus_scan_worker_spec.rb]: Very minor: What does `type: :worker` do and do we definitely need it?

File: spec/workers/virus_scan_worker_spec.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  let(:worker) { described_class.new }
```
Comment (floehopper) [inline spec/workers/virus_scan_worker_spec.rb]: Very minor: I might be wrong, but I think you might be able to achieve this more succinctly with `subject(:worker)`.

File: spec/workers/virus_scan_worker_spec.rb [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  let(:asset) { FactoryGirl.create(:asset) }
+
+  it "calls out to the VirusScanner to scan the file" do
+    scanner = double("VirusScanner")
+    expect(VirusScanner).to receive(:new).with(asset.file.path).and_return(scanner)
```
Comment (floehopper) [inline spec/workers/virus_scan_worker_spec.rb]: Minor: I would move the first of these lines into a `let(:scanner)` statement and use an `instance_double` of `VirusScanner`. And I'd move the second line into a `before` block and convert the `expect` into an `allow`.

Comment (floehopper) [reply spec/workers/virus_scan_worker_spec.rb]: I've just realised that you've moved these examples from the `Asset` spec, so you can ignore my previous comment and I will refrain from making more suggestions for improvement in this commit.

File: app/models/asset.rb [hunk 5]
```diff
@@ -135,7 +135,7 @@ def reset_state
   end
 
   def schedule_virus_scan
-    self.delay.scan_for_viruses if self.unscanned?
+    VirusScanWorker.perform_async(self.id) if self.unscanned?
```
Comment (floehopper) [inline app/models/asset.rb]: Minor: I know you mentioned it in a previous commit note, but I think it would be nice to mention in the note for this commit that you're leaving `scan_for_viruses` executing the virus scanning synchronously to support any remaining Delayed Jobs in the queue.

File: spec/models/asset_spec.rb [hunk 6]
```diff
@@ -226,43 +226,38 @@
...
     it "does not schedule a scan after update if the file is unchanged" do
       a = FactoryGirl.create(:clean_asset)
       a.created_at = 5.days.ago
-      expect {
-        a.save!
-      }.not_to change(Delayed::Job, :count)
+
+      expect(VirusScanWorker).not_to receive(:perform_async).with(a)
+
+      a.save!
```
Comment (floehopper) [inline spec/models/asset_spec.rb]: Minor: I wonder whether this means we can stop including `DelayedJobHelpers` in this spec?

File: spec/requests/virus_scanning_spec.rb [hunk 7]
```diff
@@ -56,7 +56,7 @@ def initialize
     get "/media/#{asset.id}/eicar.com"
     expect(response).to have_http_status(:not_found)
 
-    run_all_delayed_jobs
+    VirusScanWorker.drain
```
Comment (floehopper) [inline spec/requests/virus_scanning_spec.rb]: Minor: I wonder whether this means we can stop automatically including the `DelayedJobHelpers` module and maybe even delete it? Maybe we can do that as a separate PR in which we remove all DJ related code...?

File: spec/models/asset_spec.rb [hunk 8]
```diff
@@ -226,43 +226,38 @@
...
     end
 
     it "does not schedule a scan after update if the file is unchanged" do
       a = FactoryGirl.create(:clean_asset)
       a.created_at = 5.days.ago
-      expect {
-        a.save!
-      }.not_to change(Delayed::Job, :count)
+
+      expect(VirusScanWorker).not_to receive(:perform_async).with(a)
```
Comment (floehopper) [inline spec/models/asset_spec.rb]: I think the `with` should be passed `a.id` vs `a`. Having said that, I think you could drop the `with` altogether here.

File: spec/requests/virus_scanning_spec.rb (hunk 7 above)
Comment (chrislo) [reply spec/requests/virus_scanning_spec.rb]: I was thinking of removing all of the delayed job code in one PR - but yes, it does leave things a bit messy until then.

File: .rubocop_todo.yml (hunk 1 above)
Comment (chrislo) [reply .rubocop_todo.yml]: Yes, there's still unrelated violations for these exceptions even with the virus related blocks moved across. I'll see if I can address those in a separate PR.

File: spec/workers/virus_scan_worker_spec.rb (hunk 2 above)
Comment (chrislo) [reply spec/workers/virus_scan_worker_spec.rb]: No - it was added by the sidekiq rspec generator, I'll remove it.

File: spec/workers/virus_scan_worker_spec.rb (hunk 3 above)
Comment (chrislo) [reply spec/workers/virus_scan_worker_spec.rb]: Rspec supports [implicit subject](https://relishapp.com/rspec/rspec-core/v/2-0/docs/subject/implicit-subject) but using them to make this more succinct violates [RuboCop::Cop::RSpec::NamedSubject](http://www.rubydoc.info/gems/rubocop-rspec/1.6.0/RuboCop/Cop/RSpec/NamedSubject). There's not much difference between `let(:worker) { described_class.new }` and `subject(:worker) { described_class.new }` so I'll leave it.

File: app/models/asset.rb (hunk 5 above)
Comment (chrislo) [reply app/models/asset.rb]: Thanks! I'll add a note to the commit message.

File: spec/models/asset_spec.rb (hunk 8 above)
Comment (chrislo) [reply spec/models/asset_spec.rb]: Good catch - I'll fix that. Thanks!

**Comment by chrislo** [comment]
Thanks for the review @floehopper. I've addressed your feedback by rebase and am force-pushing before merging.
